Import data:

In [1]:
import pandas as pd

data = pd.read_csv('./new-york-city-airbnb-open-data/AB_NYC_2019.csv')
data.head(10)

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   
5  5099         Large Cozy 1 BR Apartment In Midtown East     7322   
6  5121                                   BlissArtsSpace!     7356   
7  5178                  Large Furnished Room Near B'way      8967   
8  5203                Cozy Clean Guest Room - Family Apt     7490   
9  5238                Cute & Cozy Lower East Side 1 bdrm     7549   

     host_name neighbourhood_group       neighbourhood  latitude  longitude  \
0         John            Brooklyn          Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan             Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan              Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn        Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan         East Harlem  40.79851  -73.94399   
5        Chris           Manhattan         Murray Hill  40.74767  -73.97500   
6        Garon            Brooklyn  Bedford-Stuyvesant  40.68688  -73.95596   
7     Shunichi           Manhattan      Hell's Kitchen  40.76489  -73.98493   
8    MaryEllen           Manhattan     Upper West Side  40.80178  -73.96723   
9          Ben           Manhattan           Chinatown  40.71344  -73.99037   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   
5  Entire home/apt    200               3                 74  2019-06-22   
6     Private room     60              45                 49  2017-10-05   
7     Private room     79               2                430  2019-06-24   
8     Private room     79               2                118  2017-07-21   
9  Entire home/apt    150               1                160  2019-06-09   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0  
5               0.59                               1               129  
6               0.40                               1                 0  
7               3.47                               1               220  
8               0.99                               1                 0  
9               1.33                               4               188

Check non-null counts and data type of columns: 

In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

I can see that there are 37457 unique hosts out of 48879:

In [3]:
len(data['host_id'].unique())

37457

For each host_id the number of unique host names is 1 (or host name is unknown):

In [4]:
filt = data.groupby("host_id").host_name.nunique() != 1
data.groupby("host_id").host_name.first()[filt]

host_id
415290       None
526653       None
919218       None
5162530      None
5300585      None
7779204      None
7822683      None
23077718     None
24576978     None
26138712     None
32722063     None
33134899     None
39608626     None
100971588    None
119609345    None
159156636    None
177146433    None
228750026    None
Name: host_name, dtype: object

For each host_id the number of unique calculated_host_listings_count is 1:

In [5]:
(data.groupby("host_id").calculated_host_listings_count.nunique() == 1).all()

True

Those 3 columns represent information specific to hosts. 
So I can create hosts_list representing hosts data:

In [96]:
hosts_list = data[['host_id', 'host_name', 'calculated_host_listings_count']]
hosts_list

host_id      host_name  calculated_host_listings_count
0          2787           John                               6
1          2845       Jennifer                               2
2          4632      Elisabeth                               1
3          4869    LisaRoxanne                               1
4          7192          Laura                               1
...         ...            ...                             ...
48890   8232441        Sabrina                               2
48891   6570630        Marisol                               2
48892  23492952  Ilgar & Aysel                               1
48893  30985759            Taz                               6
48894  68119814     Christophe                               1

[48895 rows x 3 columns]

Drop duplicate rows in hosts_list and set host_id as index:

In [97]:
pd.options.mode.chained_assignment = None
hosts_list.drop_duplicates(inplace=True)
hosts_list.rename(columns={"host_id": "id"}, inplace=True)
hosts_list.set_index("id", inplace=True)
hosts_list

host_name  calculated_host_listings_count
id                                                      
2787                John                               6
2845            Jennifer                               2
4632           Elisabeth                               1
4869         LisaRoxanne                               1
7192               Laura                               1
...                  ...                             ...
274307600       Jonathan                               1
274311461          Scott                               1
274321313            Kat                               1
23492952   Ilgar & Aysel                               1
68119814      Christophe                               1

[37457 rows x 2 columns]

There are 221 unique neighbourhood values:

In [51]:
len(data["neighbourhood"].unique())

221

For each neighbourhood the number of unique neighbourhood_group is 1:

In [7]:
(data.groupby("neighbourhood").neighbourhood_group.nunique() == 1).all()

True

Those 2 columns represent information specific to neighbourhood location. I can create neighbourhood_list representing neighbourhood data. Then I drop duplicate rows nad set index name as id: 

In [55]:
neighbourhood_list = data[["neighbourhood", "neighbourhood_group"]]
neighbourhood_list.drop_duplicates(inplace=True)
neighbourhood_list.reset_index(inplace=True)
neighbourhood_list.drop(columns=["index"], inplace=True)
neighbourhood_list.index.name = "id"
neighbourhood_list

neighbourhood neighbourhood_group
id                                   
0      Kensington            Brooklyn
1         Midtown           Manhattan
2          Harlem           Manhattan
3    Clinton Hill            Brooklyn
4     East Harlem           Manhattan
..            ...                 ...
216   Bull's Head       Staten Island
217      New Dorp       Staten Island
218     Rossville       Staten Island
219  Breezy Point              Queens
220   Willowbrook       Staten Island

[221 rows x 2 columns]

Other columns in data are related to BnB places:

In [69]:
places_list = data.drop(columns=["host_name", "calculated_host_listings_count", "neighbourhood_group"])
places_list.set_index("id", inplace=True)
places_list

name   host_id  \
id                                                                      
2539                     Clean & quiet apt home by the park      2787   
2595                                  Skylit Midtown Castle      2845   
3647                    THE VILLAGE OF HARLEM....NEW YORK !      4632   
3831                        Cozy Entire Floor of Brownstone      4869   
5022       Entire Apt: Spacious Studio/Loft by central park      7192   
...                                                     ...       ...   
36484665    Charming one bedroom - newly renovated rowhouse   8232441   
36485057      Affordable room in Bushwick/East Williamsburg   6570630   
36485431            Sunny Studio at Historical Neighborhood  23492952   
36485609               43rd St. Time Square-cozy single bed  30985759   
36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

               neighbourhood  latitude  longitude        room_type  price  \
id                                                                          
2539              Kensington  40.64749  -73.97237     Private room    149   
2595                 Midtown  40.75362  -73.98377  Entire home/apt    225   
3647                  Harlem  40.80902  -73.94190     Private room    150   
3831            Clinton Hill  40.68514  -73.95976  Entire home/apt     89   
5022             East Harlem  40.79851  -73.94399  Entire home/apt     80   
...                      ...       ...        ...              ...    ...   
36484665  Bedford-Stuyvesant  40.67853  -73.94995     Private room     70   
36485057            Bushwick  40.70184  -73.93317     Private room     40   
36485431              Harlem  40.81475  -73.94867  Entire home/apt    115   
36485609      Hell's Kitchen  40.75751  -73.99112      Shared room     55   
36487245      Hell's Kitchen  40.76404  -73.98933     Private room     90   

          minimum_nights  number_of_reviews last_review  reviews_per_month  \
id                                                                           
2539                   1                  9  2018-10-19               0.21   
2595                   1                 45  2019-05-21               0.38   
3647                   3                  0         NaN                NaN   
3831                   1                270  2019-07-05               4.64   
5022                  10                  9  2018-11-19               0.10   
...                  ...                ...         ...                ...   
36484665               2                  0         NaN                NaN   
36485057               4                  0         NaN                NaN   
36485431              10                  0         NaN                NaN   
36485609               1                  0         NaN                NaN   
36487245               7                  0         NaN                NaN   

          availability_365  
id                          
2539                   365  
2595                   355  
3647                   365  
3831                   194  
5022                     0  
...                    ...  
36484665                 9  
36485057                36  
36485431                27  
36485609                 2  
36487245                23  

[48895 rows x 12 columns]

Change neighbourhood column to corresponding ids:

In [70]:
places_list.rename(columns={"neighbourhood": "neighbourhood_id"}, inplace=True)
places_list["neighbourhood_id"] = places_list["neighbourhood_id"].map(lambda x: neighbourhood_list[neighbourhood_list["neighbourhood"] == x].index.values.astype(int)[0])
places_list

name   host_id  \
id                                                                      
2539                     Clean & quiet apt home by the park      2787   
2595                                  Skylit Midtown Castle      2845   
3647                    THE VILLAGE OF HARLEM....NEW YORK !      4632   
3831                        Cozy Entire Floor of Brownstone      4869   
5022       Entire Apt: Spacious Studio/Loft by central park      7192   
...                                                     ...       ...   
36484665    Charming one bedroom - newly renovated rowhouse   8232441   
36485057      Affordable room in Bushwick/East Williamsburg   6570630   
36485431            Sunny Studio at Historical Neighborhood  23492952   
36485609               43rd St. Time Square-cozy single bed  30985759   
36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

          neighbourhood_id  latitude  longitude        room_type  price  \
id                                                                        
2539                     0  40.64749  -73.97237     Private room    149   
2595                     1  40.75362  -73.98377  Entire home/apt    225   
3647                     2  40.80902  -73.94190     Private room    150   
3831                     3  40.68514  -73.95976  Entire home/apt     89   
5022                     4  40.79851  -73.94399  Entire home/apt     80   
...                    ...       ...        ...              ...    ...   
36484665                 6  40.67853  -73.94995     Private room     70   
36485057                21  40.70184  -73.93317     Private room     40   
36485431                 2  40.81475  -73.94867  Entire home/apt    115   
36485609                 7  40.75751  -73.99112      Shared room     55   
36487245                 7  40.76404  -73.98933     Private room     90   

          minimum_nights  number_of_reviews last_review  reviews_per_month  \
id                                                                           
2539                   1                  9  2018-10-19               0.21   
2595                   1                 45  2019-05-21               0.38   
3647                   3                  0         NaN                NaN   
3831                   1                270  2019-07-05               4.64   
5022                  10                  9  2018-11-19               0.10   
...                  ...                ...         ...                ...   
36484665               2                  0         NaN                NaN   
36485057               4                  0         NaN                NaN   
36485431              10                  0         NaN                NaN   
36485609               1                  0         NaN                NaN   
36487245               7                  0         NaN                NaN   

          availability_365  
id                          
2539                   365  
2595                   355  
3647                   365  
3831                   194  
5022                     0  
...                    ...  
36484665                 9  
36485057                36  
36485431                27  
36485609                 2  
36487245                23  

[48895 rows x 12 columns]

Connect to database:

In [100]:
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base

postgresql_password = os.environ.get('POSTGRES_PASSWORD')
db_string = f"postgresql://postgres:{postgresql_password}@127.0.0.1:5432/bnb_places"
db = create_engine(db_string)
Base = declarative_base()

Create database structure:

In [91]:
from sqlalchemy import Column, Integer, String, Float, Date, ForeignKey, CheckConstraint, UniqueConstraint

class Place(Base):
    __tablename__ = 'places'
    __table_args__ = (
        CheckConstraint('price >= 0'),
        CheckConstraint('minimum_nights >= 0'),
        CheckConstraint('number_of_reviews >= 0'),
        CheckConstraint('reviews_per_month >= 0'),
        CheckConstraint('availability_365 >= 0 AND availability_365 <= 365')
    )
    id = Column(Integer, primary_key = True)
    name = Column(String(500), nullable = True)
    host_id = Column(Integer, ForeignKey('hosts.id'))
    latitude = Column(Float, nullable=False)
    longitude = Column(Float, nullable=False)
    neighbourhood_id = Column(Integer, ForeignKey('neighbourhoods.id'))
    room_type = Column(String(30), nullable = False)
    price = Column(Float, nullable=False)
    minimum_nights = Column(Integer, nullable=False)
    number_of_reviews = Column(Integer, nullable=False)
    last_review = Column(Date, nullable=True)
    reviews_per_month = Column(Float, nullable=True)
    availability_365 = Column(Integer, nullable=False)

class Host(Base):
    __tablename__ = 'hosts'
    __table_args__ = (
        CheckConstraint('calculated_host_listings_count >= 0'),
    )
    id = Column(Integer, primary_key=True)
    host_name = Column(String(200), nullable=True)
    calculated_host_listings_count = Column(Integer, nullable=False)
    
class Neighbourhood(Base):
    __tablename__ = 'neighbourhoods'
    id = Column(Integer, primary_key=True )
    neighbourhood = Column(String(100), nullable=False)
    neighbourhood_group = Column(String(100), nullable=False)

Create all tables:

In [92]:

Base.metadata.create_all(db)

Insert data:

In [99]:
hosts_list.to_sql('hosts', db, if_exists='append')
neighbourhood_list.to_sql('neighbourhoods', db, if_exists='append')
places_list.to_sql('places', db, if_exists='append')